WangscapeTest/TestCartesianPower.cpp

```diff
@@ -0,0 +1,199 @@
+#include <gtest/gtest.h>
+
+#include <vector>
+#include <algorithm>
+
+#include <tilegen/CartesianPower.h>
+
+class TestCartesianPower : public ::testing::Test
+{
+protected:
+    typedef std::vector<int> Container;
+    typedef std::vector<int>::const_iterator Iterator;
+
+    Container base{4, 5, 6};
+    const tilegen::CartesianPower<Iterator> cp{base, 4};
+};
+
+TEST_F(TestCartesianPower, TestCartesianPowerSize)
+{
+    EXPECT_EQ(81, cp.size()) <<
+        "CartesianProduct has incorrect size";
+}
+
+TEST_F(TestCartesianPower, TestCartesianPowerCount)
+{
+    size_t count = 0;
+    for (const auto& it : cp)
+        count++;
+    EXPECT_EQ(81, count);
+}
+
+TEST_F(TestCartesianPower, TestCartesianPowerComparisonBeginEnd)
+{
+    EXPECT_EQ(cp.cbegin(), cp.cbegin());
+    EXPECT_NE(cp.cend(), cp.cbegin());
+    EXPECT_NE(cp.cbegin(), cp.cend());
+    EXPECT_EQ(cp.cend(), cp.cend());
+
+    EXPECT_EQ(cp.cbegin(), cp.begin());
+    EXPECT_EQ(cp.cend(), cp.end());
+}
+
+TEST_F(TestCartesianPower, TestCartesianPowerComparisonLast)
+{
+    auto it = cp.cbegin();
+    for (size_t i = 0; i < cp.size(); i++)
+        ++it;
+    EXPECT_EQ(it, it);
+    EXPECT_EQ(it, cp.cend());
+    EXPECT_EQ(cp.cend(), it);
+    EXPECT_NE(it, cp.cbegin());
+    EXPECT_NE(cp.cbegin(), it);
+}
+
+TEST_F(TestCartesianPower, TestCartesianPowerComparisonOther)
+{
+    tilegen::CartesianPower<Iterator> cp1{base.cbegin() + 1, 2, 4};
+    tilegen::CartesianPower<Iterator> cp2{base.cbegin(), base.size(), 3};
+    EXPECT_NE(cp.begin(), cp1.begin());
+    EXPECT_NE(cp.begin(), cp2.begin());
+}
+
+TEST_F(TestCartesianPower, TestCartesianPowerComparisonOtherMid)
+{
+    auto it = cp.cbegin();
+    std::advance(it, 23);
+    EXPECT_EQ(it, it);
+    EXPECT_NE(it, cp.cbegin());
+    EXPECT_NE(cp.cbegin(), it);
+    EXPECT_NE(it, cp.cend());
+    EXPECT_NE(cp.cend(), it);
+}
+
+TEST_F(TestCartesianPower, TestCartesianPowerIteratorCopyable)
+{
+    static_assert(std::is_copy_constructible<tilegen::CartesianPowerIterator<Iterator>>::value,
+                  "CartesianPowerIterator not CopyConstructible");
+    static_assert(std::is_copy_assignable<tilegen::CartesianPowerIterator<Iterator>>::value,
+                  "CartesianPowerIterator not CopyAssignable");
+    auto it = base.cbegin();
+    // try to copy-construct
+    auto it2 = it;
+    // try to copy-assign
+    it = base.cend();
```

Comment:
Is this test needed? It hasn't any runtime assert. Could these `static_assert`s be placed "around" `CartesianPowerIterator` class? Or it seems that there could be `EXPECT_EQ`, checking if assignment was succesful.